File: Python/openstk/sys/typex.py
from __future__ import annotations
import itertools
from .mscorlib.System import Reflection, Type, attributex, newx, memberx

def RAssembly(name: str = None):
    def clscall(cls):
        nonlocal name
        def attribute_(self, obj): self.ltypes = obj.ltypes if hasattr(obj, 'ltypes') else None
        attributex(cls, 'RAssembly', name, attribute_)
        def funccall(func):
            def wrapper(*args, **kwargs): print(f'wrapper: {func.__name__} for {cls.__name__}'); return func(*args, **kwargs)
            return wrapper
        return cls
    return clscall

def RType(name: str = None, valueType: bool = False):
    def clscall(cls):
        nonlocal name
        attributex(cls, 'RType', f'{cls.__module__[cls.__module__.index('formats.')+8:]}.{cls.__qualname__.replace('+' if name != '+' else '\00', '.')}' if name == None or name == '+' else name, TypeX.getRType)
        def funccall(func):
            def wrapper(*args, **kwargs): print(f'wrapper: {func.__name__} for {cls.__name__}'); return func(*args, **kwargs)
            return wrapper
        return cls
    return clscall

# AssemblyTag
class AssemblyTag:
    def __init__(self, ltypes: dict[str, dict[str, type]], rtypes: dict[str, type]):
        self.ltypes: dict[str, dict[str, type]] = ltypes
        self.rtypes: dict[str, type] = rtypes

# TypeX
class TypeX:
    assemblys: dict[str, module] = { 'mscorlib': Reflection.Assembly.getAssembly(Type) }
    assemblyRedirects: dict[str, module] = {}
    tags: dict[str, AssemblyTag] = {}

    @staticmethod
    def getRType(source: module, typeName: str, throwOnError: bool = True, rtype: bool = False, obj: object = None) -> type:
        # print(f'getRType{"R" if rtype else ""}: {source.__name__ if source else ""} - {typeName}')
        @staticmethod
        def _assemblyResolver(assembly):
            # print(f'_assemblyResolver: {assembly.name}')
            if assembly.name in TypeX.assemblys and (a := TypeX.assemblys[assembly.name]): return a
            a = TypeX.assemblys[assembly.name] = s if assembly.name in TypeX.assemblyRedirects and (s := TypeX.assemblyRedirects[assembly.name]) else Reflection.Assembly.load(assembly.fullName)
            return a
        @staticmethod
        def _typeResolver(assembly, name, throwOnError):
            nonlocal source
            a = assembly or source
            # print(f'_typeResolver: {a.__name__} - {name}')
            if a.__name__ in TypeX.tags and (tag := TypeX.tags[a.__name__]):
                # r-types
                if name in tag.rtypes and (type := tag.rtypes[name]): return type
                # l-types
                idx = name.rfind('.')
                ns, na = (name[:idx], name[(idx + 1):]) if idx != -1 else (None, None)
                if ns and ns in tag.ltypes and (b := tag.ltypes[ns]) and na in b and (type := b[na]): return type
            return a.getType(name, throwOnError)
        return Type.getType(typeName, assemblyResolver=_assemblyResolver, typeResolver=_typeResolver, throwOnError=throwOnError, rtype=rtype, obj=obj)

    @staticmethod
    def scanTypes(types: list[type]) -> None:
        for type in types:
            if type.__module__ in TypeX.tags: continue
            assembly = Reflection.Assembly.getAssembly(type)
            attribs = assembly._attribs_ if hasattr(assembly, '_attribs_') else []
            rassembiles = [(s.name, s.ltypes) for s in attribs if s.kind == 'RAssembly']
            rtypes = [(s.type, s.name) for s in attribs if s.kind == 'RType']
            assemblyRedirects = {s[0]:assembly for s in rassembiles if s[0]}
            if len(assemblyRedirects) > 0: TypeX.assemblyRedirects |= assemblyRedirects
            ltypes_ = {i:ss[i] for ss in [s[1] for s in rassembiles if s[1]] for i in ss} # ..GroupBy(s => s.Key).ToDictionary(s => s.Key, s => s.First().Value),
            # for k, g in itertools.groupby(sorted(ltypes_, key=lambda x: x.type), key=lambda x: x.type): v = next(g); readersByType[v.type] = v
            TypeX.tags[assembly.__name__] = AssemblyTag(
                ltypes=ltypes_,
                rtypes={s[1]:s[0] for s in rtypes})

    @staticmethod
    def getDefaultConstructor(type: type) -> callable: return type #return lambda: newx(type)
    @staticmethod
    def getAllProperties(type: type) -> list: return memberx(type)._properties if hasattr(type, '_properties') else []
    @staticmethod
    def getAllFields(type: type) -> list: return memberx(type)._fields if hasattr(type, '_fields') else []

File: dotnet/Core/OpenStk.Poly/TypeX.cs
using System.Collections.Generic;
using System.Linq;
using System.Reflection;

namespace System;

[AttributeUsage(AttributeTargets.Class | AttributeTargets.Enum)]
public class RAssemblyAttribute(string name = null) : Attribute {
    public string Name = name;
    internal Dictionary<string, Dictionary<string, Type>> GetLTypes(Type cls) => (Dictionary<string, Dictionary<string, Type>>)cls.GetField("LTypes")?.GetValue(null);
}

[AttributeUsage(AttributeTargets.Class | AttributeTargets.Enum)]
public class RTypeAttribute(string name = null) : Attribute {
    public string Name = name;
    internal string GetName(Type cls) => Name == null || Name == "+"
        ? cls.FullName[(cls.FullName.IndexOf("Formats.") + 8)..].Replace(Name != "+" ? "+" : "\x00", ".")
        : Name;
}

// AssemblyTag
class AssemblyTag {
    public Dictionary<string, Dictionary<string, Type>> LTypes;
    public Dictionary<string, Type> RTypes;
}

// TypeX
public static class TypeX {
    static readonly Dictionary<string, Assembly> Assemblys = [];
    static Dictionary<string, Assembly> AssemblyRedirects = [];
    static readonly Dictionary<Assembly, AssemblyTag> Tags = [];

    public static void ScanTypes(Type[] types) {
        foreach (var type in types) {
            if (Tags.ContainsKey(type.Assembly)) continue;
            var assembly = type.Assembly;
            var attribs = assembly.GetTypes();
            var rassembiles = attribs.SelectMany(s => Attribute.GetCustomAttributes(s, typeof(RAssemblyAttribute)).Cast<RAssemblyAttribute>().Select(t => (t.Name, LTypes: t.GetLTypes(s))));
            var rtypes = attribs.SelectMany(s => Attribute.GetCustomAttributes(s, typeof(RTypeAttribute)).Cast<RTypeAttribute>().Select(t => (Type: s, Name: t.GetName(s))));
            var assemblyRedirects = rassembiles.Where(s => s.Name != null).ToDictionary(s => s.Name, s => assembly);
            if (assemblyRedirects.Count > 0) AssemblyRedirects = AssemblyRedirects.Concat(assemblyRedirects).ToDictionary(s => s.Key, s => s.Value);
            Tags[assembly] = new AssemblyTag {
                LTypes = rassembiles.Where(s => s.LTypes != null).SelectMany(s => s.LTypes).GroupBy(s => s.Key).ToDictionary(s => s.Key, s => s.First().Value),
                RTypes = rtypes.ToDictionary(s => s.Name, s => s.Type),
            };
        }
    }

    public static Type GetRType(this Assembly source, string typeName, bool throwOnError = true) => Type.GetType(typeName,
        assemblyResolver: assembly =>
            Assemblys.TryGetValue(assembly.Name, out var a) ? a
            : Assemblys[assembly.FullName] = AssemblyRedirects.TryGetValue(assembly.Name, out var s) ? s : Assembly.Load(assembly.FullName),
        typeResolver: (assembly, name, throwOnError) => {
            var a = assembly ?? source;
            if (Tags.TryGetValue(a, out var tag)) {
                // r-types
                if (tag.RTypes.TryGetValue(name, out var type)) return type;
                // l-types
                var idx = name.LastIndexOf('.');
                var (ns, na) = idx != -1 ? (name[..idx], name[(idx + 1)..]) : default;
                if (ns != null && tag.LTypes.TryGetValue(ns, out var b) && b.TryGetValue(na, out type)) return type;
            }
            return a.GetType(name, throwOnError);
        }, throwOnError);

    public static ConstructorInfo GetDefaultConstructor(this Type type) => type.GetConstructor(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance, null, [], null);
    public static PropertyInfo[] GetAllProperties(this Type type) => [.. type.GetProperties(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly).ToList().FindAll(p => p.GetGetMethod(true) != null && p.GetGetMethod(true) == p.GetGetMethod(true).GetBaseDefinition())];
    public static FieldInfo[] GetAllFields(this Type type) => type.GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly);
}